### PR TITLE
Enable changing executor params through governance

### LINKED ISSUE #6805: Executor environment parameters should be configurable on the fly
Currently, executor environment parameters may only be changed with the runtime upgrade. We need the ability to configure them through governance.

### PR DESCRIPTION
**Pull request by s0me0ne-unkn0wn**
Closes #6805 

Add a pallet call to `session_info` which sets new executor parameters. The new parameter set is stored as a separate storage value and becomes effective starting from the next session.

### CODE REVIEW

**Comment by s0me0ne-unkn0wn** [comment]
Currently, I'm not sure I'm doing the right thing. Feel free to guide me on the right path.

File: runtime/parachains/src/session_info.rs [hunk 1]
```diff
@@ -113,6 +108,25 @@ pub mod pallet {
...
+	/// Pending executor parameters
+	#[pallet::storage]
+	pub(crate) type PendingExecutorParams<T: Config> =
+		StorageValue<_, Option<ExecutorParams>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((
+			Weight::from_parts(0, 0), // FIXME: How do we calculate weights?
```
Comment (s0me0ne-unkn0wn) [inline runtime/parachains/src/session_info.rs:121]: Where does the weight come from?

File: runtime/parachains/src/session_info.rs [hunk 2]
```diff
@@ -196,10 +210,20 @@ impl<T: Config> Pallet<T> {
 			dispute_period,
 		};
 		Sessions::<T>::insert(&new_session_index, &new_session_info);
-		SessionExecutorParams::<T>::insert(
-			&new_session_index,
-			ExecutorParams::from(&EXECUTOR_PARAMS[..]),
-		);
+
+		let new_executor_params = if let Some(executor_params) = <PendingExecutorParams<T>>::get() {
```
Comment (s0me0ne-unkn0wn) [inline runtime/parachains/src/session_info.rs:214]: Should I add a migration to put `None` to the `PendingExecutorParams` for the first time?

File: runtime/parachains/src/session_info.rs [hunk 3]
```diff
@@ -196,10 +210,20 @@ impl<T: Config> Pallet<T> {
...
+			<PendingExecutorParams<T>>::put(&None);
+			executor_params
+		} else {
+			let current_session_index = <shared::Pallet<T>>::session_index();
```
Comment (s0me0ne-unkn0wn) [inline runtime/parachains/src/session_info.rs:218]: Is it okay to request the current session index from inside of the `initializer_on_new_session` handler? Wouldn't it return a new session index? If yes, then where can I get the old one? Just `.saturated_sub(1)` from the new one maybe?

File: runtime/parachains/src/session_info.rs [hunk 4]
```diff
@@ -113,6 +108,25 @@ pub mod pallet {
...
+		StorageValue<_, Option<ExecutorParams>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((
+			Weight::from_parts(0, 0), // FIXME: How do we calculate weights?
+			DispatchClass::Operational,
+		))]
+		pub fn set_executor_params(origin: OriginFor<T>, new: ExecutorParams) -> DispatchResult {
```
Comment (s0me0ne-unkn0wn) [inline runtime/parachains/src/session_info.rs:124]: Is adding this call enough to allow governance to control the executor params?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2564962

File: runtime/parachains/src/session_info.rs (hunk 2 above)
Comment (slumber) [reply runtime/parachains/src/session_info.rs:214]: No

File: runtime/parachains/src/session_info.rs (hunk 4 above)
Comment (slumber) [reply runtime/parachains/src/session_info.rs:124]: Yes, should be

File: runtime/parachains/src/session_info.rs (hunk 3 above)
Comment (slumber) [reply runtime/parachains/src/session_info.rs:218]: `initializer_on_new_session` is first called for `shared`, then for everyone else.

For safety you could change the `SessionChangeNotification` to include `prev_session_index` returned by `shared`, I guess.

File: runtime/parachains/src/session_info.rs (hunk 1 above)
Comment (slumber) [reply runtime/parachains/src/session_info.rs:121]: https://github.com/paritytech/polkadot/pull/4938 try to follow this pr for example, or feel free to ask if you're lost

File: runtime/parachains/src/session_info.rs (hunk 3 above)
Comment (rphmeier) [reply runtime/parachains/src/session_info.rs:218]: > Is it okay to request the current session index from inside of the initializer_on_new_session handler? Wouldn't it return a new session index? If yes, then where can I get the old one? Just .saturated_sub(1) from the new one maybe?

Yeah, saturated_sub seems fine. The point of having the `initializer_on_new_session` approach is to have a clear view into relative initialization order of pallets used for parachains, so we can safely do this kind of thing.

Comment (slumber) [reply runtime/parachains/src/session_info.rs:218]: nit: this index is available in the same function above `let new_session_index...`
Reading it from storage again is unnecessary.

**Review by rphmeier — COMMENTED** [review]
Is there a reason why the executor params can't just go into the host configuration via the `Configuration` pallet?

**Comment by s0me0ne-unkn0wn** [comment]
> Is there a reason why the executor params can't just go into the host configuration via the `Configuration` pallet?

I believe there are no technical obstacles, but I have two concerns about that:
1) Currently, the `configuration` only contains values of primitive types, like `u32` and `BlockNumber`. Is it okay to bring the whole `ExecutorParams` with all its derivatives to it?
2) Do `ExecutorParams` fit the `HostConfiguration` concept at all?

Beyond that, I have no other arguments against moving them to the `configuration` pallet altogether (but storage migration will be needed then).

File: runtime/parachains/src/session_info.rs [hunk 5]
```diff
@@ -113,6 +108,25 @@ pub mod pallet {
 	#[pallet::getter(fn session_executor_params)]
 	pub(crate) type SessionExecutorParams<T: Config> =
 		StorageMap<_, Identity, SessionIndex, ExecutorParams>;
+
+	/// Pending executor parameters
+	#[pallet::storage]
+	pub(crate) type PendingExecutorParams<T: Config> =
+		StorageValue<_, Option<ExecutorParams>, ValueQuery>;
```
Comment (ggwpez) [inline runtime/parachains/src/session_info.rs:115]:
```suggestion
		StorageValue<_, ExecutorParams, OptionQuery>;
```

File: runtime/parachains/src/session_info.rs [hunk 6]
```diff
@@ -113,6 +108,25 @@ pub mod pallet {
...
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((
+			Weight::from_parts(0, 0), // FIXME: How do we calculate weights?
+			DispatchClass::Operational,
+		))]
+		pub fn set_executor_params(origin: OriginFor<T>, new: ExecutorParams) -> DispatchResult {
+			ensure_root(origin)?;
+			<PendingExecutorParams<T>>::put(Some(new));
```
Comment (ggwpez) [inline runtime/parachains/src/session_info.rs:126]: And probably emit an event, so wallets and indexers can be notified.

File: runtime/parachains/src/session_info.rs [hunk 7]
```diff
@@ -196,10 +210,20 @@ impl<T: Config> Pallet<T> {
...
+
+		let new_executor_params = if let Some(executor_params) = <PendingExecutorParams<T>>::get() {
+			<PendingExecutorParams<T>>::put(&None);
+			executor_params
+		} else {
+			let current_session_index = <shared::Pallet<T>>::session_index();
+			if let Some(executor_params) = <SessionExecutorParams<T>>::get(current_session_index) {
+				executor_params
+			} else {
+				unreachable!("Current session executor params must always be present");
```
Comment (ggwpez) [inline runtime/parachains/src/session_info.rs:222]: There is no advantage here of having a potential panic (which must be avoided in consensus code at all costs, see [SE](https://substrate.stackexchange.com/questions/317/when-to-use-expect)).  

Rather use `frame_support::defensive!("Current session executor params must always be present")` and skip the insertion below, which should incur no harm - besides the pending params upgrade being discarded.

**Comment by bkchr** [comment]
> To me it feels like `HostConfiguration` has become a [God object](https://en.wikipedia.org/wiki/God_object) of sorts where many unrelated knobs are being put.

I think the main problem here being that we always need to decode the entire `HostConfiguration` for just accessing some small detail of the configuration. Maybe with the `ExecutorParams` it is a good start to put these into their own `Configuration` storage item. Nevertheless, the configuration pallet should be re-used to host this.

File: runtime/parachains/src/session_info.rs (hunk 6 above)
Comment (s0me0ne-unkn0wn) [reply runtime/parachains/src/session_info.rs:126]: Do they need to know about execution parameters change? That's a validators' business iiuc.

**Comment by s0me0ne-unkn0wn** [comment]
@bkchr, one more concern is that we're accessing executor params on the session base, not on the state base, and we're maintaining a sliding window of different executor params for the sessions inside the dispute period. The `session_info` pallet maintains some infrastructure to support that (similar to `SessionInfo` storage), like keeping the earliest stored session number, etc. Moving the whole thing to the `configuration` pallet would introduce calls from the `configuration` pallet to the storage of the `session_info` pallet. Are we okay with that?

**Comment by bkchr** [comment]
> etc. Moving the whole thing to the `configuration` pallet would introduce calls from the `configuration` pallet to the storage of the `session_info` pallet. Are we okay with that?

Isn't the session info not just caching certain parameters per session and also currently caches things that are already part of the host configuration?

**Comment by slumber** [comment]
@s0me0ne-unkn0wn configuration change is always buffered to be applied in a couple of sessions, if this is somethings you're worried about

**Comment by s0me0ne-unkn0wn** [comment]
> > 
> 
> Isn't the session info not just caching certain parameters per session and also currently caches things that are already part of the host configuration?

In principle, yes. But
1) It maintains a sliding window of those hashes to make them available when referred by session index, not by relay parent
2) It works the other way round: `session_info` gets some information from `configuration` and then caches it, and that's quite natural. But I'm not sure how natural it is for the `configuration` to request from `session_info` (I didn't check yet, but at least I'm afraid to get a pallet initialization order conflict)

**Comment by s0me0ne-unkn0wn** [comment]
Moved pending changes to the `configuration` pallet. Now the whole thing looks much more logical and straightforward to me.

**Comment by mrcnski** [comment]
> Moved pending changes to the `configuration` pallet. Now the whole thing looks much more logical and straightforward to me.

+1. To me it definitely makes sense to have executor params reflected in the
session info, as we need to keep them per session. But the pending value
should be present in some external configuration which feeds the session info,
and this other configuration can be governanced.

**Comment by slumber** [comment]
> Moved pending changes to the `configuration` pallet. Now the whole thing looks much more logical and straightforward to me.

I believe your code should respect
https://github.com/paritytech/polkadot/blob/1fe1702508f24b96b136f696fc180d6b27145c8b/runtime/parachains/src/shared.rs#L33-L36

**Comment by eskimor** [comment]
> > To me it feels like `HostConfiguration` has become a [God object](https://en.wikipedia.org/wiki/God_object) of sorts where many unrelated knobs are being put.
> 
> I think the main problem here being that we always need to decode the entire `HostConfiguration` for just accessing some small detail of the configuration. Maybe with the `ExecutorParams` it is a good start to put these into their own `Configuration` storage item. Nevertheless, the configuration pallet should be re-used to host this.

But that should be cached - right? Or I guess it is cached encoded? Also each read costs, so grouping things together should improve performance (assuming we are reading multiple values in the same block).

Anyhow, my main concern about factoring this out is code duplication / and or varying behavior breaking expectations. E.g. that 2 session delay for applying changes. Also what is the long term plan - make each configuration value its own storage value? Or do we still want some grouping?

File: runtime/parachains/src/configuration.rs [hunk 8]
```diff
@@ -1163,6 +1170,21 @@ pub mod pallet {
...
 			})
 		}
+
+		/// Set executor parameters for the next session
+		#[pallet::call_index(46)]
+		#[pallet::weight((
+			Weight::from_parts(0, 0), // FIXME: Benchmark weights
+			DispatchClass::Operational,
+		))]
+		pub fn set_next_session_executor_params(
```
Comment (eskimor) [inline runtime/parachains/src/configuration.rs:1180]: As mentioned, I am not really a fan of breaking consistency here. If we want to get rid of that god object, we should have some over arching plan for it. Right now we just have a god object and some random stuff extra, which even has different semantics when it comes to the timing of taking effect. I would just put it in the god object and leave splitting that one up for another PR.

Comment (s0me0ne-unkn0wn) [reply runtime/parachains/src/configuration.rs:1180]: A good point! The whole thing should be refactored first before we'll be able to tear it apart.
Then the next question is, what part of `HostConfiguration` does it belong to? We have three:
* The parameters that are required for the parachains;
* The parameters that are not essential, but still may be of interest for parachains;
* Parameters that will unlikely be needed by parachains.

I bet on the second one, as we were discussing the collators would need it. That discussion is somewhat stalled, but that's still a concern.

Comment (eskimor) [reply runtime/parachains/src/configuration.rs:1180]: Looking at how current values are grouped, it feels a bit arbitrary. E.g. max code size is mandatory, but max pov size is not?

Comment (slumber) [reply runtime/parachains/src/configuration.rs:1180]: > Looking at how current values are grouped, it feels a bit arbitrary. E.g. max code size is mandatory, but max pov size is not?

Grouping basically means that parachains can decode the [top](https://github.com/paritytech/polkadot/blob/26ff2cc11a6d9f7776c5ed7376f89c37fd214fbe/primitives/src/v4/mod.rs#L1062-L1066) configuration. Groups 2 and 3 don't really make difference to me. For now only 1 cannot be changed without special treatment.

And max pov size is available through validation data.

**Comment by s0me0ne-unkn0wn** [comment]
I'm a little bit stuck with the `configuration` pallet migration.

If I add something to `HostConfiguration` I have to add a migration that translates storage. But `HostConfiguation` is not a versioned primitive, and we have an [active storage migration from `v4` to `v5`](https://github.com/paritytech/polkadot/blob/173a5a325177dd63ae32b55e11a35df5fb29163e/runtime/parachains/src/configuration/migration.rs#L34). It refers to the [current version of `HostConfiguration`](https://github.com/paritytech/polkadot/blob/173a5a325177dd63ae32b55e11a35df5fb29163e/runtime/parachains/src/configuration/migration.rs#L188) as a translation target structure, so if I change anything in `HostConfiguration`, I have to change the migration.

And here I'm not sure, if that migration to `v5` is over now, and the code may be safely removed and replaced with similar migration from `v5` to `v6`, or if I should add my changes to the existing migration from `v4` to `v5`. How do I determine that?

**Comment by eskimor** [comment]
> I'm a little bit stuck with the `configuration` pallet migration.
> 
> If I add something to `HostConfiguration` I have to add a migration that translates storage. But `HostConfiguation` is not a versioned primitive, and we have an [active storage migration from `v4` to `v5`](https://github.com/paritytech/polkadot/blob/173a5a325177dd63ae32b55e11a35df5fb29163e/runtime/parachains/src/configuration/migration.rs#L34). It refers to the [current version of `HostConfiguration`](https://github.com/paritytech/polkadot/blob/173a5a325177dd63ae32b55e11a35df5fb29163e/runtime/parachains/src/configuration/migration.rs#L188) as a translation target structure, so if I change anything in `HostConfiguration`, I have to change the migration.
> 
> And here I'm not sure, if that migration to `v5` is over now, and the code may be safely removed and replaced with similar migration from `v5` to `v6`, or if I should add my changes to the existing migration from `v4` to `v5`. How do I determine that?

If the v5 migration has been released already, we have to do a v6. Not sure how well this works with runtime releases being skipped, but it is the only sensible way.

**Comment by s0me0ne-unkn0wn** [comment]
> If the v5 migration has been released already

I've checked that the latest release branch (`release-v0.9.41`) does not include the v5 migration, so it's not been released yet? How to know for sure?

**Comment by slumber** [comment]
> > If the v5 migration has been released already
> 
> I've checked that the latest release branch (`release-v0.9.41`) does not include the v5 migration, so it's not been released yet? How to know for sure?

@coderobe

**Comment by eskimor** [comment]
> > If the v5 migration has been released already
> 
> I've checked that the latest release branch (`release-v0.9.41`) does not include the v5 migration, so it's not been released yet? How to know for sure?

That's how you are sure. If it is not yet on the latest release branch, It cannot be released. I mean in theory things could be backported, would still be very weird (and a bug) to not have that backport in the next release as well.

**Comment by ordian** [comment]
I think the idea was to keep previous migrations for a few releases at least. Since each migration does version check (`if on_chain_version == expected`), it's safe to keep outdated migrations. Whereas, if we remove them prematurely, it will cause problems for us, potentially corrupting the storage.

**Comment by s0me0ne-unkn0wn** [comment]
> Whereas, if we remove them prematurely, it will cause problems for us, potentially corrupting the storage.

I'm much more concerned about if it is safe to add my changes to the existing migration :) If it's not been released, then it's safe, right?

File: runtime/parachains/src/configuration.rs [hunk 9]
```diff
@@ -1163,6 +1170,21 @@ pub mod pallet {
 				config.async_backing_params = new;
 			})
 		}
+
+		/// Set executor parameters for the next session
+		#[pallet::call_index(46)]
+		#[pallet::weight((
+			Weight::from_parts(0, 0), // FIXME: Benchmark weights
```
[the remark this answers is not in the document]
Comment (s0me0ne-unkn0wn) [reply runtime/parachains/src/configuration.rs:1173]: Benchmarks are nearly ready in #7033. What do I do next? Just merge that branch into this one?
Other weights have also changed a little, that's not a problem iiuc?

**Review by ordian — APPROVED** [review]

**Comment by s0me0ne-unkn0wn** [comment]
bot merge
